File: java/PathFind.java
package path_find;

import java.util.ArrayList;

public class PathFind {
    private final ArrayList<ArrayList<Integer>> arr;
    private final Cord startDot;
    private final Cord finishDot;

    private static final int[][] directions = {
            {0, 1},
            {1, 0},
            {-1, 0},
            {0, -1}
    };

    private final ArrayList<Cord> reachable;
    private final ArrayList<Cord> explored;

    PathFind(ArrayList<ArrayList<Integer>> arr, Cord startDot, Cord finishDot) {
        this.arr = arr;
        this.startDot = startDot;
        this.finishDot = finishDot;

        this.reachable = new ArrayList<>();
        this.reachable.add(this.startDot);

        this.explored = new ArrayList<>();
    }

    public ArrayList<Cord> findPath() {
        while (!this.reachable.isEmpty()) {
            Cord next_step = this.reachable.get(0);

            if (next_step.x == this.finishDot.x && next_step.y == this.finishDot.y) {
                this.finishDot.prevCord = next_step.prevCord;

                return this.buildPath();
            }

            // remove from reachable next_step
            for (int i = 0; i < this.reachable.size(); i++) {
                Cord iterable = this.reachable.get(i);

                if (iterable.x == next_step.x && iterable.y == next_step.y) {
                    this.reachable.remove(i);
                    break;
                }
            }

            //add to explored arr
            this.explored.add(next_step);

            for (int[] direction : directions) {
                int newX = next_step.x + direction[0];
                int newY = next_step.y + direction[1];

                if (newX >= 0 && newX < this.arr.size() && newY >= 0 && newY < this.arr.get(0).size()) {
                    int possibleMove = this.arr.get(newX).get(newY);

                    Cord newReachable = new Cord(newX, newY, next_step);

                    boolean cordIsNotInReachable = PathFind.cordIsNotInArr(newReachable, this.reachable);
                    boolean cordIsNotInExplored = PathFind.cordIsNotInArr(newReachable, this.explored);
                    boolean isPossibleMove = possibleMove == 0;

                    if (cordIsNotInReachable && cordIsNotInExplored && isPossibleMove) {
                        this.reachable.add(newReachable);
                    }
                }
            }
        }

        return new ArrayList<>();
    }

    private static boolean cordIsNotInArr(Cord cord, ArrayList<Cord> arr) {
        if (!arr.isEmpty()) {
            for (Cord value : arr) {
                if (value.x == cord.x && value.y == cord.y) {
                    return false;
                }
            }
        }

        return true;
    }

    private ArrayList<Cord> buildPath() {
        ArrayList<Cord> path = new ArrayList<>();

        Cord step = this.finishDot;

        while (step.prevCord.x != -1 && step.prevCord.y != -1) {
            path.add(step);

            step = step.prevCord;
        }

        path.add(this.startDot);

        // reverse array list
        ArrayList<Cord> reversedPath = new ArrayList<>();

        for (int i = path.size() - 1; i >= 0; i--) {

            // Append the elements in reverse order
            reversedPath.add(path.get(i));
        }

        return reversedPath;
    }
}

File: c#/PathFind.cs
namespace path_find;

public class PathFind
{
    private List<List<int>> arr;

    private Cord _startDot;
    private Cord _finishDot;

    private static readonly List<int[]> Directions = new List<int[]>()
    {
        new int[2] { 0, 1 },
        new int[2] { 1, 0 },
        new int[2] { -1, 0 },
        new int[2] { 0, -1 }
    };

    private List<Cord> _reachable;
    private List<Cord> _explored;

    public PathFind(List<List<int>> arr, Cord startDot, Cord finishDot)
    {
        this.arr = arr;
        _startDot = startDot;
        _finishDot = finishDot;

        _reachable = new List<Cord>();
        _explored = new List<Cord>();

        _reachable.Add(_startDot);
    }

    private static bool IsCordInArr(Cord cord, List<Cord> arr)
    {
        foreach (var t in arr)
        {
            if (t.x == cord.x && t.y == cord.y)
            {
                return true;
            }
        }

        return false;
    }

    public List<Cord> FindPath()
    {
        while (_reachable.Count > 0)
        {
            var nextStep = _reachable[0];

            if (nextStep.x == _finishDot.x && nextStep.y == _finishDot.y)
            {
                _finishDot.prevCord = nextStep.prevCord;

                return BuildPath();
            }

            for (var i = 0; i < _reachable.Count; i++)
            {
                if (_reachable[i].x != nextStep.x || _reachable[i].y != nextStep.y) continue;
                _reachable.RemoveAt(i);

                break;
            }

            _explored.Add(nextStep);

            foreach (var direction in Directions)
            {
                var newX = nextStep.x + direction[0];
                var newY = nextStep.y + direction[1];

                if (newX >= 0 && newX < arr.Count && newY >= 0 && newY < arr[0].Count)
                {
                    var possibleMove = arr[newX][newY];

                    var newReachable = new Cord(newX, newY, nextStep);

                    var cordIsNotInReachable = !IsCordInArr(newReachable, _reachable);
                    var cordIsNotInExplored = !IsCordInArr(newReachable, _explored);
                    var isPossibleMove = possibleMove == 0;

                    if (cordIsNotInReachable && cordIsNotInExplored && isPossibleMove)
                    {
                        _reachable.Add(newReachable);
                    }
                }
            }
        }

        return new List<Cord>();
    }

    private List<Cord> BuildPath()
    {
        var path = new List<Cord>();

        var step = _finishDot;

        while (step.prevCord != null)
        {
            path.Add(step);

            step = step.prevCord;
        }

        path.Add(_startDot);

        path.Reverse();

        return path;
    }
}